### PR TITLE
Implement efficient real/complex 1D even-length transforms

### PR DESCRIPTION
**Pull request by malcolmroberts**
Summary of proposed changes:
-  Implement time and memory efficient real/complex 1D transforms for even problem sizes.
-  Includes some general code cleaning.

### CODE REVIEW

File: clients/tests/hipfft_test.cpp [hunk 1]
```diff
@@ -89,7 +103,14 @@ TEST(hipfftTest, CheckBufferSizeD2Z)
...
 
     EXPECT_TRUE(hipfftMakePlan1d(plan, n, HIPFFT_D2Z, batch, &workSize) == HIPFFT_SUCCESS);
-    EXPECT_TRUE(workSize == 2 * n * batch * sizeof(double));
+    if(n % 2 == 0)
+    {
+        EXPECT_TRUE(workSize == 0);
+    }
+    else
+    {
+        EXPECT_TRUE(workSize == 2 * n * sizeof(float));
```
Comment (feizheng10) [inline clients/tests/hipfft_test.cpp]: double?

File: clients/tests/hipfft_test.cpp [hunk 2]
```diff
@@ -103,7 +124,14 @@ TEST(hipfftTest, CheckBufferSizeZ2D)
...
 
     EXPECT_TRUE(hipfftMakePlan1d(plan, n, HIPFFT_Z2D, batch, &workSize) == HIPFFT_SUCCESS);
-    EXPECT_TRUE(workSize == 2 * n * batch * sizeof(double));
+    if(n % 2 == 0)
+    {
+        EXPECT_TRUE(workSize == 0);
+    }
+    else
+    {
+        EXPECT_TRUE(workSize == 2 * n * sizeof(float));
```
Comment (feizheng10) [inline clients/tests/hipfft_test.cpp]: double?

File: clients/tests/hipfft_test.cpp (hunk 1 above)
Comment (malcolmroberts) [reply clients/tests/hipfft_test.cpp]: Thanks; fixed in 0cfc23b.

File: clients/tests/hipfft_test.cpp (hunk 2 above)
Comment (malcolmroberts) [reply clients/tests/hipfft_test.cpp]: Thanks; fixed in 0cfc23b.

File: library/src/plan.cpp [hunk 3]
```diff
@@ -192,6 +265,12 @@ rocfft_status rocfft_plan_create_internal(rocfft_plan                   plan,
             {
                 if(description->inArrayType != description->outArrayType)
                     return rocfft_status_invalid_array_type;
+                if(description->inStrides != description->outStrides)
```
Comment (feizheng10) [inline library/src/plan.cpp]: double check this, I thought we removed it.

Comment (malcolmroberts) [reply library/src/plan.cpp]: The type has been changed from size_t* to std::array<size_t, 3>, so the check is on the data, not the address.  However, this isn't the correct place to check this, so we're probably just looking at the default values, so I removed it in d91b984.

File: library/src/plan.cpp [hunk 4]
```diff
@@ -1744,99 +1944,119 @@ void TreeNode::TraverseTreeAssignBuffersLogicA(OperatingBuffer& flipIn,
...
+                if(parent->iOffset != 0)
+                    iOffset = parent->iOffset;
+                break;
+            default:
+                inArrayType = rocfft_array_type_complex_interleaved;
+            }
         }
 
-        switch(obOut)
+        if(outArrayType == rocfft_array_type_unset)
```
Comment (feizheng10) [inline library/src/plan.cpp]: we might remove redundant code if the logic for in/out are the same later.

**Comment by akilaMD** [comment]
So this is still failing. I say we merge it in for now and deal with the librocfft linking issues ASAP.

File: library/include/rocfft.h [hunk 5]
```diff
@@ -182,10 +183,10 @@ ROCFFT_EXPORT rocfft_status rocfft_plan_create(rocfft_plan*                  pla
...
  * rocfft_execution_info_create
  *  */
-ROCFFT_EXPORT rocfft_status rocfft_execute(const rocfft_plan     plan,
-                                           void*                 in_buffer[],
-                                           void*                 out_buffer[],
-                                           rocfft_execution_info info);
+ROCFFT_EXPORT rocfft_status rocfft_execute(const rocfft_plan           plan,
+                                           void*                       in_buffer[],
+                                           void*                       out_buffer[],
+                                           const rocfft_execution_info info);
```
Comment (bragadeesh) [inline library/include/rocfft.h]: info cannot be const, I think it was intended to act as both an input & output parameter, the library may be returning information back to user

File: library/include/rocfft.h [hunk 6]
```diff
@@ -210,54 +211,48 @@ ROCFFT_EXPORT rocfft_status rocfft_plan_description_set_scale_float( rocfft_plan
 ROCFFT_EXPORT rocfft_status rocfft_plan_description_set_scale_double( rocfft_plan_description description, double scale );
```
Comment (bragadeesh) [inline library/include/rocfft.h]: change the scale to const here as well to be consistent?

File: library/include/rocfft.h (hunk 5 above)
Comment (malcolmroberts) [reply library/include/rocfft.h]: Sounds good; done.

File: library/include/rocfft.h (hunk 6 above)
Comment (malcolmroberts) [reply library/include/rocfft.h]: Sounds good; done.

**Pull request merged**